Module resolution is broken when wildcard path exists

### Versions
```
    _                      _                 ____ _     ___
   / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
  / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
 / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
/_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
               |___/
    
Angular CLI: 1.7.4
Node: 9.9.0
OS: darwin x64
Angular: 5.2.10
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

```

### Repro steps

Add the following to tsconfig.json:

1. Create a new CLI project:

`ng new angular-cli-is-broken`

2. Add a wildcard path mapping to the tsconfig:

```js
"paths": {
  "*": ["./src/app/*"]
}
```

3. Build the project:

```
npm run build
```

### Observed behavior

Resolution of modules does not work:

```
ERROR in ./src/app/app.module.ngfactory.js
Module not found: Error: Can't resolve '/Users/jon/GitHub/angular-cli-is-broken/src/src/app/@angular/common' in '/Users/jon/GitHub/angular-cli-is-broken/src/app'
```

Etc.


### Desired behavior

It should resolve modules correctly, as it did in 1.6.

### Mention any other details that might be useful (optional)

This worked perfectly in 1.6. You can confirm this by regressing the CLI version in package.json to 1.6, and confirming that `ng build` now works.

--- a/packages/ngtools/webpack/package.json
+++ b/packages/ngtools/webpack/package.json
@@ -22,6 +22,7 @@
   "homepage": "https://github.com/angular/angular-cli/tree/master/packages/@ngtools/webpack",
   "dependencies": {
     "@angular-devkit/core": "0.0.0",
+    "enhanced-resolve": "4.1.0",    
     "rxjs": "6.2.2",
     "tree-kill": "1.2.0",
     "webpack-sources": "1.2.0"

--- a/packages/ngtools/webpack/src/angular_compiler_plugin.ts
+++ b/packages/ngtools/webpack/src/angular_compiler_plugin.ts
@@ -32,7 +32,7 @@ import { WebpackCompilerHost, workaroundResolve } from './compiler_host';
 import { resolveEntryModuleFromMain } from './entry_resolver';
 import { gatherDiagnostics, hasErrors } from './gather_diagnostics';
 import { LazyRouteMap, findLazyRoutes } from './lazy_routes';
-import { resolveWithPaths } from './paths-plugin';
+import { TypeScriptPathsPlugin } from './paths-plugin';
 import { WebpackResourceLoader } from './resource_loader';
 import {
   exportLazyModuleMap,
@@ -729,6 +729,14 @@ export class AngularCompilerPlugin {
     });
 
     compiler.hooks.afterResolvers.tap('angular-compiler', compiler => {
+      // tslint:disable-next-line:no-any
+      (compiler as any).resolverFactory.hooks.resolver
+        .for('normal')
+        // tslint:disable-next-line:no-any
+        .tap('angular-compiler', (resolver: any) => {
+          new TypeScriptPathsPlugin(this._compilerOptions).apply(resolver);
+        });
+
       compiler.hooks.normalModuleFactory.tap('angular-compiler', nmf => {
         // Virtual file system.
         // TODO: consider if it's better to remove this plugin and instead make it wait on the
@@ -753,21 +761,6 @@ export class AngularCompilerPlugin {
           },
         );
       });
-    });
-
-    compiler.hooks.normalModuleFactory.tap('angular-compiler', nmf => {
-      nmf.hooks.beforeResolve.tapAsync(
-        'angular-compiler',
-        (request: NormalModuleFactoryRequest, callback: Callback<NormalModuleFactoryRequest>) => {
-          resolveWithPaths(
-            request,
-            callback,
-            this._compilerOptions,
-            this._compilerHost,
-            this._moduleResolutionCache,
-          );
-        },
-      );
     });
   }
 

--- a/packages/ngtools/webpack/src/paths-plugin.ts
+++ b/packages/ngtools/webpack/src/paths-plugin.ts
@@ -6,54 +6,97 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import * as path from 'path';
-import * as ts from 'typescript';
-import {
-  Callback,
-  NormalModuleFactoryRequest,
-} from './webpack';
+import { CompilerOptions, MapLike } from 'typescript';
+import { NormalModuleFactoryRequest } from './webpack';
 
+const getInnerRequest = require('enhanced-resolve/lib/getInnerRequest');
 
-export function resolveWithPaths(
-  request: NormalModuleFactoryRequest,
-  callback: Callback<NormalModuleFactoryRequest>,
-  compilerOptions: ts.CompilerOptions,
-  host: ts.CompilerHost,
-  cache?: ts.ModuleResolutionCache,
-) {
-  if (!request || !request.request || !compilerOptions.paths) {
-    callback(null, request);
+export interface TypeScriptPathsPluginOptions extends Pick<CompilerOptions, 'paths' | 'baseUrl'> {
 
-    return;
+}
+
+export class TypeScriptPathsPlugin {
+  constructor(private _options: TypeScriptPathsPluginOptions) { }
+
+  // tslint:disable-next-line:no-any
+  apply(resolver: any) {
+    if (!this._options.paths || Object.keys(this._options.paths).length === 0) {
+      return;
+    }
+
+    const target = resolver.ensureHook('resolve');
+    const resolveAsync = (request: NormalModuleFactoryRequest, requestContext: {}) => {
+      return new Promise<NormalModuleFactoryRequest | undefined>((resolve, reject) => {
+        resolver.doResolve(
+          target,
+          request,
+          '',
+          requestContext,
+          (error: Error | null, result: NormalModuleFactoryRequest | undefined) => {
+            if (error) {
+              reject(error);
+            } else {
+              resolve(result);
+            }
+          },
+        );
+      });
+    };
+
+    resolver.getHook('described-resolve').tapPromise(
+      'TypeScriptPathsPlugin',
+      async (request: NormalModuleFactoryRequest, resolveContext: {}) => {
+        if (!request || request.typescriptPathMapped) {
+          return;
+        }
+
+        const originalRequest = getInnerRequest(resolver, request);
+        if (!originalRequest) {
+          return;
+        }
+
+        // Only work on Javascript/TypeScript issuers.
+        if (!request.context.issuer || !request.context.issuer.match(/\.[jt]sx?$/)) {
+          return;
+        }
+
+        // Relative or absolute requests are not mapped
+        if (originalRequest.startsWith('.') || originalRequest.startsWith('/')) {
+          return;
+        }
+
+        // Amd requests are not mapped
+        if (originalRequest.startsWith('!!webpack amd')) {
+          return;
+        }
+
+        const replacements = findReplacements(originalRequest, this._options.paths || {});
+        for (const potential of replacements) {
+          const potentialRequest = {
+            ...request,
+            request: path.resolve(this._options.baseUrl || '', potential),
+            typescriptPathMapped: true,
+          };
+          const result = await resolveAsync(potentialRequest, resolveContext);
+
+          if (result) {
+            return result;
+          }
+        }
+      },
+    );
   }
+}
 
-  // Only work on Javascript/TypeScript issuers.
-  if (!request.contextInfo.issuer || !request.contextInfo.issuer.match(/\.[jt]sx?$/)) {
-    callback(null, request);
-
-    return;
-  }
-
-  const originalRequest = request.request.trim();
-
-  // Relative requests are not mapped
-  if (originalRequest.startsWith('.') || originalRequest.startsWith('/')) {
-    callback(null, request);
-
-    return;
-  }
-
-  // Amd requests are not mapped
-  if (originalRequest.startsWith('!!webpack amd')) {
-    callback(null, request);
-
-    return;
-  }
-
+function findReplacements(
+  originalRequest: string,
+  paths: MapLike<string[]>,
+): Iterable<string> {
   // check if any path mapping rules are relevant
   const pathMapOptions = [];
-  for (const pattern in compilerOptions.paths) {
+  for (const pattern in paths) {
     // get potentials and remove duplicates; JS Set maintains insertion order
-    const potentials = Array.from(new Set(compilerOptions.paths[pattern]));
+    const potentials = Array.from(new Set(paths[pattern]));
     if (potentials.length === 0) {
       // no potential replacements so skip
       continue;
@@ -96,9 +139,7 @@ export function resolveWithPaths(
   }
 
   if (pathMapOptions.length === 0) {
-    callback(null, request);
-
-    return;
+    return [];
   }
 
   // exact matches take priority then largest prefix match
@@ -112,63 +153,23 @@ export function resolveWithPaths(
     }
   });
 
-  if (pathMapOptions[0].potentials.length === 1) {
-    const onlyPotential = pathMapOptions[0].potentials[0];
-    let replacement;
-    const starIndex = onlyPotential.indexOf('*');
-    if (starIndex === -1) {
-      replacement = onlyPotential;
-    } else if (starIndex === onlyPotential.length - 1) {
-      replacement = onlyPotential.slice(0, -1) + pathMapOptions[0].partial;
-    } else {
-      const [prefix, suffix] = onlyPotential.split('*');
-      replacement = prefix + pathMapOptions[0].partial + suffix;
-    }
-
-    request.request = path.resolve(compilerOptions.baseUrl || '', replacement);
-    callback(null, request);
-
-    return;
-  }
-
-  // TODO: The following is used when there is more than one potential and will not be
-  //       needed once this is turned into a full webpack resolver plugin
-
-  const moduleResolver = ts.resolveModuleName(
-    originalRequest,
-    request.contextInfo.issuer,
-    compilerOptions,
-    host,
-    cache,
-  );
-
-  const moduleFilePath = moduleResolver.resolvedModule
-                         && moduleResolver.resolvedModule.resolvedFileName;
-
-  // If there is no result, let webpack try to resolve
-  if (!moduleFilePath) {
-    callback(null, request);
-
-    return;
-  }
-
-  // If TypeScript gives us a `.d.ts`, it is probably a node module
-  if (moduleFilePath.endsWith('.d.ts')) {
-    // If in a package, let webpack resolve the package
-    const packageRootPath = path.join(path.dirname(moduleFilePath), 'package.json');
-    if (!host.fileExists(packageRootPath)) {
-      // Otherwise, if there is a file with a .js extension use that
-      const jsFilePath = moduleFilePath.slice(0, -5) + '.js';
-      if (host.fileExists(jsFilePath)) {
-        request.request = jsFilePath;
+  const replacements: string[] = [];
+  pathMapOptions.forEach(option => {
+    for (const potential of option.potentials) {
+      let replacement;
+      const starIndex = potential.indexOf('*');
+      if (starIndex === -1) {
+        replacement = potential;
+      } else if (starIndex === potential.length - 1) {
+        replacement = potential.slice(0, -1) + option.partial;
+      } else {
+        const [prefix, suffix] = potential.split('*');
+        replacement = prefix + option.partial + suffix;
       }
+
+      replacements.push(replacement);
     }
+  });
 
-    callback(null, request);
-
-    return;
-  }
-
-  request.request = moduleFilePath;
-  callback(null, request);
+  return replacements;
 }

--- a/packages/ngtools/webpack/src/webpack.ts
+++ b/packages/ngtools/webpack/src/webpack.ts
@@ -16,7 +16,9 @@ export interface Callback<T = any> {
 
 export interface NormalModuleFactoryRequest {
   request: string;
+  context: { issuer: string };
   contextInfo: { issuer: string };
+  typescriptPathMapped?: boolean;
 }
 
 export interface InputFileSystem {

--- a/tests/legacy-cli/e2e/tests/misc/module-resolution.ts
+++ b/tests/legacy-cli/e2e/tests/misc/module-resolution.ts
@@ -73,14 +73,14 @@ export default async function () {
 
   await updateJsonFile('tsconfig.json', tsconfig => {
     tsconfig.compilerOptions.paths = {
-      '@firebase/polyfill': ['@firebase/polyfill/index.ts'],
+      '@firebase/polyfill': ['./node_modules/@firebase/polyfill/index.ts'],
     };
   });
   await expectToFail(() => ng('build'));
 
   await updateJsonFile('tsconfig.json', tsconfig => {
     tsconfig.compilerOptions.paths = {
-      '@firebase/polyfill*': ['@firebase/polyfill/index.ts'],
+      '@firebase/polyfill*': ['./node_modules/@firebase/polyfill/index.ts'],
     };
   });
   await expectToFail(() => ng('build'));

